hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java

```diff
@@ -1399,23 +1408,75 @@ private void createMissingCFsInMetaDuringUpgrade(TableDescriptor metaDescriptor)
+   * @param ri Region info.
    * @return True if region is online and scannable else false if an error or shutdown (Otherwise we
    *         just block in here holding up all forward-progess).
+   * @throws IOException If the master restart is required.
    */
-  private boolean isRegionOnline(RegionInfo ri) {
+  private boolean isRegionOnline(RegionInfo ri) throws IOException {
     RetryCounter rc = null;
     while (!isStopped()) {
       RegionState rs = this.assignmentManager.getRegionStates().getRegionState(ri);
       if (rs != null && rs.isOpened()) {
         if (this.getServerManager().isServerOnline(rs.getServerName())) {
           return true;
+        } else {
+          ServerName serverNameForRegion = rs.getServerName();
+          Optional<Procedure<MasterProcedureEnv>> scpForServer =
+            this.procedureExecutor.getProcedures().stream()
+              .filter(p -> p instanceof ServerCrashProcedure
+                && serverNameForRegion.equals(((ServerCrashProcedure) p).getServerName()))
+              .findFirst();
+          if (!scpForServer.isPresent()) {
+            LOG.info("{} has state {} but the server {} is not online, scheduling recovery.",
+              ri.getRegionNameAsString(), rs, rs.getServerName());
+            this.getServerManager().expireServer(rs.getServerName(), true);
+            int numRetries = this.getConfiguration()
+              .getInt(HBASE_MASTER_REGION_SCHEDULE_RECOVERY_WAIT_RETRIES, 20);
+            int sleepInterval = this.getConfiguration()
+              .getInt(HBASE_MASTER_REGION_SCHEDULE_RECOVERY_WAIT_INTERVAL_MS, 2000);
+            while (numRetries > 0) {
+              scpForServer = this.procedureExecutor.getProcedures().stream()
+                .filter(p -> p instanceof ServerCrashProcedure
+                  && serverNameForRegion.equals(((ServerCrashProcedure) p).getServerName()))
+                .findFirst();
+              if (scpForServer.isPresent()) {
+                ServerCrashProcedure proc = (ServerCrashProcedure) scpForServer.get();
+                if (proc.isFinished() || proc.isSuccess()) {
+                  rs = this.assignmentManager.getRegionStates().getRegionState(ri);
+                  if (rs != null && rs.isOpened()) {
+                    if (this.getServerManager().isServerOnline(rs.getServerName())) {
+                      return true;
+                    }
+                  }
+                }
+              }
+              Threads.sleep(sleepInterval);
+              numRetries--;
+            }
+            if (numRetries == 0) {
+              rs = this.assignmentManager.getRegionStates().getRegionState(ri);
+              if (rs != null && rs.isOpened()) {
+                if (this.getServerManager().isServerOnline(rs.getServerName())) {
+                  return true;
+                }
+              }
+              throw new PleaseRestartMasterException("Scheduled SCP for old server for region "
```

Comment:
So the solution here is to "crash" master once we never find an SCP for this region's RS?

Reply:
Here we crash only because even after scheduling SCP for the region (meta or namespace), we still find the region not online.